FirebaseAdmin/FirebaseAdmin/HttpErrorHandler.cs

```diff
@@ -0,0 +1,107 @@
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+
+namespace FirebaseAdmin
+{
+    /// <summary>
+    /// Base class for handling HTTP error responses.
+    /// </summary>
+    internal class HttpErrorHandler
+    {
+        private static readonly IReadOnlyDictionary<HttpStatusCode, ErrorCode> HttpErrorCodes =
+            new Dictionary<HttpStatusCode, ErrorCode>()
+            {
+                { HttpStatusCode.BadRequest, ErrorCode.InvalidArgument },
+                { HttpStatusCode.Unauthorized, ErrorCode.Unauthenticated },
+                { HttpStatusCode.Forbidden, ErrorCode.PermissionDenied },
+                { HttpStatusCode.NotFound, ErrorCode.NotFound },
+                { HttpStatusCode.Conflict, ErrorCode.Conflict },
+                { (HttpStatusCode)429, ErrorCode.ResourceExhausted },
+                { HttpStatusCode.InternalServerError, ErrorCode.Internal },
+                { HttpStatusCode.ServiceUnavailable, ErrorCode.Unavailable },
+            };
+
+        /// <summary>
+        /// Parses the given HTTP response and throws a <see cref="FirebaseException"/> if it
+        /// contains an error. Does nothing if the status code of the HTTP response indicates
+        /// success.
+        /// </summary>
+        /// <param name="response">HTTP response message to process.</param>
+        /// <param name="body">The response body read from the message.</param>
+        /// <exception cref="FirebaseException">If the response contains an error.</exception>
+        internal void ThrowIfError(HttpResponseMessage response, string body)
+        {
+            if (response.IsSuccessStatusCode)
+            {
+                return;
+            }
+
+            var info = this.ExtractErrorInfo(response, body);
+            var args = new FirebaseExceptionArgs()
+            {
+                Code = info.Code,
+                Message = info.Message,
+                HttpResponse = response,
+                ResponseBody = body,
+            };
+            throw this.CreateException(args);
+        }
+
+        protected virtual ErrorInfo ExtractErrorInfo(HttpResponseMessage response, string body)
+        {
+            ErrorCode code;
+            if (!HttpErrorCodes.TryGetValue(response.StatusCode, out code))
+            {
+                code = ErrorCode.Unknown;
+            }
+
+            var message = "Unexpected HTTP response with status: "
+                + $"{(int)response.StatusCode} ({response.StatusCode})"
+                + $"{Environment.NewLine}{body}";
+            return new ErrorInfo()
+            {
+                Code = code,
+                Message = message,
+            };
+        }
+
+        protected virtual FirebaseException CreateException(FirebaseExceptionArgs args)
+        {
+            return new FirebaseException(args.Code, args.Message, response: args.HttpResponse);
+        }
+
+        internal sealed class ErrorInfo
```

Comment:
`FirebaseException`, `ErrorInfo`, and `FirebaseExceptionArgs` all have very similar data. Could we combine a couple of these? 

Could we get rid of `ErrorInfo` and have `ExtractErrorInfo` just return a `FirebaseExceptionArgs` instead? I would say the other fields could be unset, but it looks like it has the data for the other two fields already.

Reply:
Done. Removed `ErrorInfo` altogether and using `FirebaseExceptionArgs` in its place.